File: frontend/iosApp/UILogics/Sources/UILogics/ViewModels/UserProfileViewModel.swift
import Foundation
import SwiftUI
import Combine
import Domains
import UseCases
@preconcurrency import Shared

@MainActor
public final class UserProfileViewModel: ObservableObject {
    @Published public var uiModel: UserProfileUIModel
    @Published public var selectedImage: UIImage? {
        didSet {
            if let image = selectedImage {
                pendingAvatarImage = image
            }
        }
    }
    @Published public var isShowingImagePicker = false
    @Published public var isSaving = false
    @Published public var alertItem: AlertItem?

    private var pendingAvatarImage: UIImage?
    private let useCase: UserProfileUseCaseProtocol

    public init(user: User, useCase: UserProfileUseCaseProtocol) {
        self.useCase = useCase
        self.uiModel = UserProfileUIModel(
            name: user.name,
            username: user.username,
            birthday: user.birthdayDate,
            avatarUrl: user.displayAvatarURL
        )
    }

    public func selectAvatar() {
        isShowingImagePicker = true
    }

    public func save() {
        Task {
            await saveProfile()
        }
    }

    public func showLogoutConfirmation() {
        alertItem = AlertItem(
            title: "Logout",
            message: "Are you sure you want to logout?",
            buttons: [
                Alert.Button.destructive(Text("Logout")) { [weak self] in
                    self?.useCase.logout()
                },
                Alert.Button.cancel()
            ]
        )
    }

    private func saveProfile() async {
        isSaving = true

        let avatarData: Data? = pendingAvatarImage?.jpegData(compressionQuality: 0.8)

        let result = await useCase.updateProfile(
            name: uiModel.name.trimmingCharacters(in: .whitespaces),
            birthday: uiModel.birthday,
            avatarData: avatarData
        )

        isSaving = false

        switch onEnum(of: result) {
        case .success(let success):
            uiModel.avatarUrl = success.user.displayAvatarURL
            pendingAvatarImage = nil
            selectedImage = nil
            alertItem = AlertItem(
                title: "Saved",
                message: "Your profile has been updated.",
                buttons: [Alert.Button.default(Text("OK"))]
            )
        case .successPendingSync(let pendingSync):
            uiModel.avatarUrl = pendingSync.user.displayAvatarURL
            pendingAvatarImage = nil
            selectedImage = nil
            alertItem = AlertItem(
                title: "Saved",
                message: "Your profile has been saved locally and will sync when online.",
                buttons: [Alert.Button.default(Text("OK"))]
            )
        case .failure(let failure):
            showAlert(for: failure.error)
        }
    }

    private func showAlert(for error: Shared.UpdateProfileError) {
        let message: String
        switch error {
        case .networkError:
            message = "Network error. Please check your connection and try again."
        case .serverError:
            message = "Server error. Please try again later."
        case .imageStorageFailed:
            message = "Failed to save the image locally. Please try again."
        case .databaseError:
            message = "Failed to save to local database. Please try again."
        default:
            message = "An unexpected error occurred. Please try again."
        }

        alertItem = AlertItem(
            title: "Save Failed",
            message: message,
            buttons: [Alert.Button.default(Text("OK"))]
        )
    }
}

// MARK: - UI Models

extension UserProfileViewModel {
    public struct UserProfileUIModel: Equatable {
        public var name: String
        public let username: String
        public var birthday: Date?
        public var avatarUrl: URL?

        public init(name: String, username: String, birthday: Date?, avatarUrl: URL?) {
            self.name = name
            self.username = username
            self.birthday = birthday
            self.avatarUrl = avatarUrl
        }
    }
}

File: frontend/composeApp/src/androidMain/kotlin/com/example/memoriesapp/ui/uilogics/viewmodels/UserProfileViewModel.kt
package com.example.memoriesapp.ui.uilogics.viewmodels

import android.graphics.Bitmap
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.memoriesapp.domain.User
import com.example.memoriesapp.usecase.UpdateProfileError
import com.example.memoriesapp.usecase.UpdateProfileResult
import com.example.memoriesapp.usecase.UserProfileUseCase
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import kotlinx.datetime.LocalDate
import java.io.ByteArrayOutputStream

class UserProfileViewModel(
    private val user: User,
    private val userProfileUseCase: UserProfileUseCase
) : ViewModel() {

    var name by mutableStateOf(user.name)
        private set
    var username by mutableStateOf(user.username)
        private set
    var birthday by mutableStateOf(user.birthday)
        private set
    var avatarUrl by mutableStateOf(user.displayAvatar)
        private set
    var selectedImage by mutableStateOf<Bitmap?>(null)
        private set
    var isSaving by mutableStateOf(false)
        private set
    var alertState by mutableStateOf<AlertState?>(null)
        private set

    private val _navigationEvent = MutableSharedFlow<NavigationEvent>()
    val navigationEvent: SharedFlow<NavigationEvent> = _navigationEvent.asSharedFlow()

    val isValid: Boolean
        get() = name.trim().isNotEmpty()

    fun onNameChange(newName: String) {
        name = newName
    }

    fun onBirthdayChange(newBirthday: LocalDate?) {
        birthday = newBirthday
    }

    fun onImageSelected(bitmap: Bitmap) {
        selectedImage = bitmap
    }

    fun save() {
        viewModelScope.launch {
            saveProfile()
        }
    }

    fun showLogoutConfirmation() {
        alertState = AlertState.LogoutConfirmation
    }

    fun confirmLogout() {
        viewModelScope.launch {
            userProfileUseCase.logout()
            _navigationEvent.emit(NavigationEvent.Logout)
        }
    }

    fun dismissAlert() {
        alertState = null
    }

    private suspend fun saveProfile() {
        isSaving = true

        val avatarData: ByteArray? = selectedImage?.let { bitmap ->
            val stream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, 80, stream)
            stream.toByteArray()
        }

        when (val result = userProfileUseCase.updateProfile(name.trim(), birthday, avatarData)) {
            is UpdateProfileResult.Success -> {
                name = result.user.name
                birthday = result.user.birthday
                avatarUrl = result.user.displayAvatar
                selectedImage = null
                alertState = AlertState.SaveSuccess("Your profile has been updated.", result.user)
            }
            is UpdateProfileResult.SuccessPendingSync -> {
                name = result.user.name
                birthday = result.user.birthday
                avatarUrl = result.user.displayAvatar
                selectedImage = null
                alertState = AlertState.SaveSuccess("Your profile has been saved locally and will sync when online.", result.user)
            }
            is UpdateProfileResult.Failure -> {
                alertState = AlertState.SaveError(mapError(result.error))
            }
        }

        isSaving = false
    }

    private fun mapError(error: UpdateProfileError): String {
        return when (error) {
            UpdateProfileError.NETWORK_ERROR ->
                "Network error. Please check your connection and try again."
            UpdateProfileError.SERVER_ERROR ->
                "Server error. Please try again later."
            UpdateProfileError.IMAGE_STORAGE_FAILED ->
                "Failed to save the image locally. Please try again."
            UpdateProfileError.DATABASE_ERROR ->
                "Failed to save to local database. Please try again."
            UpdateProfileError.UNKNOWN ->
                "An unexpected error occurred. Please try again."
        }
    }

    sealed class AlertState {
        data object LogoutConfirmation : AlertState()
        data class SaveSuccess(val message: String, val updatedUser: User) : AlertState()
        data class SaveError(val message: String) : AlertState()
    }

    sealed class NavigationEvent {
        data object Logout : NavigationEvent()
    }

    fun getUpdatedUser(): User {
        return User(
            id = user.id,
            name = name,
            username = username,
            birthday = birthday,
            avatarUrl = avatarUrl,
            avatarLocalPath = user.avatarLocalPath,
            syncStatus = user.syncStatus
        )
    }
}
